main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala

```diff
@@ -235,47 +234,45 @@ class CluProcessor protected (
   /** Restores the correct case for all words in a given document */
   def restoreCase(doc: Document): Unit = {
     GivenConstEmbeddingsAttachment(doc, true).perform {
-      for(sent <- doc.sentences) {
-        // the case restoration model expects lower-case words as input                    
-        val loweredWords = sent.words.map(_.toLowerCase())
+      for (sent <- doc.sentences) {
+        // The case restoration model expects lower-case words as input.
+        val originalWords = sent.words
+        val loweredWords = originalWords.map(_.toLowerCase)
         val preLabels = mtlCase.predict(0, AnnotatedSentence(loweredWords), None, getEmbeddings(doc))
         val labels = casePostProcessing(loweredWords, preLabels)
-        assert(labels.size == loweredWords.size)
-        val restoredWords = loweredWords.zip(labels).map(x => restoreCaseWord(x._1, x._2))
-        for(i <- sent.indices) {
-          sent.words(i) = restoredWords(i)
-        }
+        val restoredWords = originalWords.indices.map { index => restoreCaseWord(loweredWords(index), labels(index), originalWords(index)) }
+        restoredWords.copyToArray(originalWords)
       }
     }                    
   }
 
   private def casePostProcessing(loweredWords: IndexedSeq[String], preLabels: IndexedSeq[String]): IndexedSeq[String] = {
-    val labels = new ArrayBuffer[String]
-    for(i <- loweredWords.indices) {
-      val w = loweredWords(i)
-      for(pl <- CASE_PATTERNS) {
```

Comment:
If there are multiple `CASE_PATTERNS` that match, multiple labels will be added and the sizes of the sequences will not be equal, I think.